src/main/scala/com/ing/wbaa/airlock/proxy/provider/atlas/LineageHelpers.scala

```diff
@@ -0,0 +1,129 @@
+package com.ing.wbaa.airlock.proxy.provider.atlas
+
+import akka.http.scaladsl.model.{ ContentType, HttpRequest }
+import com.ing.wbaa.airlock.proxy.data.{ LineageGuidResponse, LineageHeaders, LineagePostGuidResponse }
+import com.ing.wbaa.airlock.proxy.provider.atlas.Model._
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.Future
+
+trait LineageHelpers extends LazyLogging with RestClient {
+
+  private def serverEntities(userSTS: String, host: String, classification: String) =
+    Entities(Seq(Server(
+      "Server",
+      userSTS,
+      ServerAttributes(host, host, host, host),
+      Seq(Classification(classification)))))
+
+  private def bucketEntities(userSTS: String, bucket: String, classification: String) =
+    Entities(Seq(Bucket(
+      "Bucket",
+      userSTS,
+      BucketAttributes(bucket, bucket, bucket),
+      Seq(Classification(classification)))))
+
+  private def fileEntities(
+      bucketGuid: String,
+      bucketObject: String,
+      userSTS: String,
+      contentType: ContentType,
+      classification: String) = {
+    Entities(Seq(IngestedFile(
+      "DataFile",
+      userSTS,
+      FileAttributes(bucketObject, bucketObject, bucketObject, contentType.mediaType.value,
+        guidRef(bucketGuid, "Bucket"),
+        Seq(Classification(classification))))))
+  }
+
+  private def processEntities(
+      serverGuid: String,
+      bucketGuid: String,
+      fileGuid: String,
+      userSTS: String,
+      method: String,
+      inputs: List[guidRef],
+      outputs: List[guidRef],
+      clientType: String,
+      timestamp: Long) = {
+    Entities(Seq(Ingestion(
+      "airlock_client",
+      userSTS,
+      IngestionAttributes(s"${clientType}_$timestamp", s"${clientType}_$timestamp", method, userSTS,
+        guidRef(serverGuid, "Server"),
+        inputs,
+        outputs))))
+  }
+
+  private def extractClient(userAgent: String): String =
+    """(\S+)/\S+""".r
+      .findFirstMatchIn(userAgent)
+      .map(_ group 1).getOrElse("generic")
+
+  private def extractHeaderOption(httpRequest: HttpRequest, header: String): Option[String] =
+    if (httpRequest.getHeader(header).isPresent)
+      Some(httpRequest.getHeader(header).get().value())
+    else
+      None
+
+  def getLineageHeaders(httpRequest: HttpRequest): LineageHeaders = {
+    val path = httpRequest.uri.path.toString().split("/")
+
+    LineageHeaders(
+      extractHeaderOption(httpRequest, "Remote-Address"),
+      path.take(path.length - 1).mkString("/"),
+      bucketObject = path.lastOption.getOrElse("emptyObject"),
+      httpRequest.method,
+      httpRequest.entity.contentType,
+      extractClient(extractHeaderOption(httpRequest, "User-Agent").getOrElse("generic")),
+      httpRequest.uri.rawQueryString,
+      extractHeaderOption(httpRequest, "x-amz-copy-source")
```

Comment:
I'd remove all getOrElse and mkString parts, this obscures a little bit that it's actually an option.
in other functions I notice you check whether these default values are set, but it'd be nice to check if the option is empty

```suggestion
      extractHeaderOption(httpRequest, "Remote-Address"),
      path.take(path.length - 1),
      bucketObject = path.lastOption,
      httpRequest.method,
      httpRequest.entity.contentType,
      extractHeaderOption(httpRequest, "User-Agent").map(extractClient),
      httpRequest.uri.rawQueryString,
      extractHeaderOption(httpRequest, "x-amz-copy-source")
```

Reply:
Suggestions on multiple lines doesn't seem to work in GitHub yet